File: samples/language-impl-comparison/scala3/pl0.scala
// PL/0 風トイ言語コンパイラ断片 (Scala 3 実装)
// PL/0 サブセットの抽象構文木とインタプリタ

package pl0

// 文
enum Stmt:
  case Assign(name: String, expr: Expr)
  case While(cond: Expr, body: List[Stmt])
  case Write(expr: Expr)

// 式
enum Expr:
  case Number(value: Int)
  case Var(name: String)
  case Binary(op: Op, lhs: Expr, rhs: Expr)

// 演算子
enum Op:
  case Add
  case Sub
  case Mul
  case Div

// ランタイム状態
case class Runtime(
  vars: Map[String, Int],
  output: List[Int]
)

// パースエラー
case class ParseError(message: String)

// 実行エラー
case class ExecError(reason: String)

// プログラムのパース (簡易実装)
def parseProgram(source: String): Either[ParseError, List[Stmt]] =
  // 実装のシンプルさを優先し、疑似実装を示す
  Right(List(
    Stmt.Assign("x", Expr.Number(10)),
    Stmt.While(
      Expr.Var("x"),
      List(
        Stmt.Write(Expr.Var("x")),
        Stmt.Assign("x", Expr.Binary(Op.Sub, Expr.Var("x"), Expr.Number(1)))
      )
    )
  ))

// 初期ランタイム状態
def initialRuntime(): Runtime = Runtime(
  vars = Map.empty,
  output = List.empty
)

// 式の評価
def evalExpr(expr: Expr, vars: Map[String, Int]): Either[ExecError, Int] =
  expr match
    case Expr.Number(n) => Right(n)
    case Expr.Var(name) =>
      vars.get(name) match
        case Some(value) => Right(value)
        case None => Left(ExecError(s"未定義変数: $name"))
    case Expr.Binary(op, lhs, rhs) =>
      for
        l <- evalExpr(lhs, vars)
        r <- evalExpr(rhs, vars)
        result <- op match
          case Op.Add => Right(l + r)
          case Op.Sub => Right(l - r)
          case Op.Mul => Right(l * r)
          case Op.Div =>
            if r == 0 then
              Left(ExecError("0で割ることはできません"))
            else
              Right(l / r)
      yield result

// 文の実行
def execStmt(stmt: Stmt, runtime: Runtime): Either[ExecError, Runtime] =
  stmt match
    case Stmt.Assign(name, expr) =>
      evalExpr(expr, runtime.vars).map { value =>
        Runtime(runtime.vars + (name -> value), runtime.output)
      }
    case Stmt.While(cond, body) =>
      execWhile(cond, body, runtime)
    case Stmt.Write(expr) =>
      evalExpr(expr, runtime.vars).map { value =>
        Runtime(runtime.vars, runtime.output :+ value)
      }

// while ループの実行
def execWhile(cond: Expr, body: List[Stmt], runtime: Runtime): Either[ExecError, Runtime] =
  def loop(current: Runtime): Either[ExecError, Runtime] =
    evalExpr(cond, current.vars) match
      case Left(err) => Left(err)
      case Right(value) =>
        if value == 0 then
          Right(current)
        else
          execStmtList(body, current) match
            case Left(err) => Left(err)
            case Right(nextState) => loop(nextState)
  loop(runtime)

// 文リストの実行
def execStmtList(stmts: List[Stmt], runtime: Runtime): Either[ExecError, Runtime] =
  stmts.foldLeft[Either[ExecError, Runtime]](Right(runtime)) { (acc, stmt) =>
    acc.flatMap(state => execStmt(stmt, state))
  }

// プログラムの実行
def exec(program: List[Stmt]): Either[ExecError, Runtime] =
  execStmtList(program, initialRuntime())

// 利用例
// parseProgram("begin x := 10; while x do write x; x := x - 1 end")
//   .flatMap(exec)
// => Right(Runtime(Map(), List(10, 9, 8, 7, 6, 5, 4, 3, 2, 1)))

File: examples/language-impl-comparison/parser-combinator/nom/pl0.rs
use nom::branch::alt;
use nom::bytes::complete::{tag, take_while1};
use nom::character::complete::{char, multispace0};
use nom::combinator::{map, map_res};
use nom::multi::separated_list1;
use nom::sequence::{delimited, pair, preceded};
use nom::IResult;

#[derive(Debug, PartialEq, Clone)]
pub enum Op {
    Add,
    Sub,
    Mul,
    Div,
}

#[derive(Debug, PartialEq, Clone)]
pub enum Expr {
    Number(i64),
    Var(String),
    Binary { op: Op, lhs: Box<Expr>, rhs: Box<Expr> },
}

#[derive(Debug, PartialEq, Clone)]
pub enum Stmt {
    Assign { name: String, expr: Expr },
    While { cond: Expr, body: Vec<Stmt> },
    Write { expr: Expr },
}

pub fn parse_program(input: &str) -> Result<Vec<Stmt>, String> {
    let parser = preceded(multispace0, block);
    match parser(input) {
        Ok((rest, stmts)) if rest.trim().is_empty() => Ok(stmts),
        Ok((_rest, _)) => Err("未消費文字が残っています".into()),
        Err(err) => Err(format!("解析に失敗しました: {err}")),
    }
}

fn block(input: &str) -> IResult<&str, Vec<Stmt>> {
    delimited(
        preceded(multispace0, tag_token("begin")),
        separated_list1(preceded(multispace0, char(';')), preceded(multispace0, stmt)),
        preceded(multispace0, tag_token("end")),
    )(input)
}

fn stmt(input: &str) -> IResult<&str, Stmt> {
    alt((while_stmt, write_stmt, assign_stmt))(input)
}

fn assign_stmt(input: &str) -> IResult<&str, Stmt> {
    map(
        pair(
            identifier,
            preceded(preceded(multispace0, tag_token(":=")), expr),
        ),
        |(name, expr)| Stmt::Assign { name, expr },
    )(input)
}

fn write_stmt(input: &str) -> IResult<&str, Stmt> {
    map(preceded(tag_token("write"), expr), |expr| Stmt::Write { expr })(input)
}

fn while_stmt(input: &str) -> IResult<&str, Stmt> {
    map(
        pair(
            preceded(tag_token("while"), expr),
            preceded(tag_token("do"), block),
        ),
        |(cond, body)| Stmt::While { cond, body },
    )(input)
}

fn expr(input: &str) -> IResult<&str, Expr> {
    parse_add_sub(input)
}

fn parse_add_sub(input: &str) -> IResult<&str, Expr> {
    let (mut input, mut node) = parse_mul_div(input)?;
    loop {
        let parsed = preceded(multispace0, alt((char('+'), char('-'))))(input);
        match parsed {
            Ok((next, op_char)) => {
                let (after_rhs, rhs) = parse_mul_div(next)?;
                let op = if op_char == '+' { Op::Add } else { Op::Sub };
                node = Expr::Binary { op, lhs: Box::new(node), rhs: Box::new(rhs) };
                input = after_rhs;
            }
            Err(_) => break,
        }
    }
    Ok((input, node))
}

fn parse_mul_div(input: &str) -> IResult<&str, Expr> {
    let (mut input, mut node) = factor(input)?;
    loop {
        let parsed = preceded(multispace0, alt((char('*'), char('/'))))(input);
        match parsed {
            Ok((next, op_char)) => {
                let (after_rhs, rhs) = factor(next)?;
                let op = if op_char == '*' { Op::Mul } else { Op::Div };
                node = Expr::Binary { op, lhs: Box::new(node), rhs: Box::new(rhs) };
                input = after_rhs;
            }
            Err(_) => break,
        }
    }
    Ok((input, node))
}

fn factor(input: &str) -> IResult<&str, Expr> {
    preceded(
        multispace0,
            alt((
                map(integer, Expr::Number),
                map(identifier, Expr::Var),
                delimited(
                    preceded(multispace0, char('(')),
                    expr,
                    preceded(multispace0, char(')')),
                ),
            )),
    )(input)
}

fn integer(input: &str) -> IResult<&str, i64> {
    map_res(preceded(multispace0, take_while1(|c: char| c.is_ascii_digit())), |digits: &str| {
        digits.parse::<i64>()
    })(input)
}

fn identifier(input: &str) -> IResult<&str, String> {
    map(preceded(multispace0, take_while1(is_ident_char)), |s: &str| s.into())(input)
}

fn is_ident_char(c: char) -> bool {
    c.is_ascii_alphanumeric() || c == '_'
}

fn tag_token<'a>(token: &'static str) -> impl Fn(&'a str) -> IResult<&'a str, &'a str> {
    move |input: &str| preceded(multispace0, tag(token))(input)
}
